runtime/src/secondary_index.rs

```diff
@@ -99,12 +103,125 @@ impl SecondaryIndexEntry for RwLockSecondaryIndexEntry {
     }
 }
 
+#[derive(Debug, Default)]
+struct HierarchicalOrderedMap<L1, L2>
+where
+    L1: Default + Sync + Send + Copy + PartialEq + Ord,
+    L2: Default + Sync + Send + Copy + PartialEq + Ord,
+{
+    capacity: usize,
+    map: Vec<(L1, L2)>,
+    excluded_keys: Vec<L1>,
+}
+
+impl<L1, L2> HierarchicalOrderedMap<L1, L2>
+where
+    L1: Default + Sync + Send + Copy + PartialEq + Ord,
+    L2: Default + Sync + Send + Copy + PartialEq + Ord,
+{
+    pub fn new(capacity: usize, excluded_keys: &[L1]) -> Self {
+        Self {
+            capacity,
+            map: Vec::new(),
+            excluded_keys: Vec::from(excluded_keys),
+        }
+    }
+    fn get_map(&self) -> &Vec<(L1, L2)> {
+        &self.map
+    }
+    fn sort_slice_by_value(&mut self, slice_key: &L1) {
+        // Obtain a slice of mutable references to all elements with the same key
+        let mut sub_slice = self
+            .map
+            .split_mut(|(k, _)| *k != *slice_key)
+            .flatten()
+            .collect_vec();
+        // Sort them
+        let mut sorting_buffer = sub_slice.iter().map(|x| *(*x)).collect_vec();
+        sorting_buffer.sort_unstable_by_key(|(_, v)| *v);
+        // Copy back into the list
+        for i in 0..sub_slice.len() {
+            *(sub_slice[i]) = sorting_buffer[i];
+        }
```

Comment:
I don't think this re-allocation and copy is necessary, is it? I would expect something like this to work:
```
sub_slice.sort_unstable_by_key(|(_, v)| *v);
```

Reply:
Would that not just sort the references in the sub-slice? Those refs represent positions in the greater list, sorting them in the sub-slice wouldn't actually change their position in the greater list, I want to actually change the data the references are pointing at, hence the temp buff to sort the actual values then copy back into the correct references positions. 

For a sanity check on my thinking I tried your recommendation. It fails my unit tests. Inspecting in the debugger confirms what I stated above, it sorts the references in the slice, but that doest actually change the order in the greater list.

Reply:
Yeah, you'll need to get rid of the flatten/collect business. Something like this should work, I think? https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=036f760622e0821965230bcb69e9cc03